zcbor/zcbor.py

```diff
@@ -2130,17 +2130,20 @@ def xcode_func_name(self):
     def repeated_xcode_func_name(self):
         return f"{self.mode}_repeated_{self.var_name(with_prefix=True)}"
 
-    def single_func_prim_name(self, union_int=None):
+    def single_func_prim_name(self, union_int=None, ptr_result=False):
         """Function name for xcoding this type, when it is a primitive type"""
+        ptr_variant = ptr_result and self.type in ["UINT", "INT", "NINT", "FLOAT", "BOOL"]
         func_prefix = self.single_func_prim_prefix()
         if self.mode == "decode":
             if self.type == "ANY":
                 func = "zcbor_any_skip"
             elif not self.is_unambiguous_value():
                 func = f"{func_prefix}_decode"
             elif not union_int:
-                func = f"{func_prefix}_expect"
+                func = f"{func_prefix}_{'p' if ptr_variant else ''}expect"
             elif union_int == "EXPECT":
+                assert not ptr_variant, \
+                       "Programmer error: expect_union should not be used like this"
```

Comment:
What about getting a little bit more formal?
```suggestion
                       "Programmer error: invalid use of expect_union."
```